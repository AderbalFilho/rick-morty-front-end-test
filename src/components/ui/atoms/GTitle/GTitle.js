import React from 'react';
import classNames from 'classnames/bind';
import PropTypes from 'prop-types';

import './GTitle.scss';

function GTitle({
  color,
  children,
  classes,
  family,
  spacing,
  tag,
  weight,
  ...rest
}) {
  const HtmlTag = tag;
  const className = classNames(
    'g-title',
    `g-title--${color}`,
    `g-title--${family}`,
    `g-title--space-${spacing}`,
    `g-title--${weight}`,
    classes
  );

  return (
    <HtmlTag className={className} {...rest}>
      {children}
    </HtmlTag>
  );
}

GTitle.defaultProps = {
  color: 'white',
  family: 'nunito',
  spacing: 'normal',
  tag: 'h1',
  weight: 'normal',
};

GTitle.propTypes = {
  children: PropTypes.node,
  classes: PropTypes.string,
  color: PropTypes.oneOf(['darkest-grey', 'light-grey', 'lime-green', 'white']),
  family: PropTypes.oneOf(['nunito', 'seravek']),
  spacing: PropTypes.oneOf(['normal', 'large']),
  tag: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  weight: PropTypes.oneOf(['normal', 'bold']),
};

export default GTitle;
