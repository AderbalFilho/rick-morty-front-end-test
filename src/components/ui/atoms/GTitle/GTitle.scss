@import 'styles/mixins';

.g-title {
  @include base-text;

  margin: 0;
  padding: 0;
}

// Color
.g-title--darkest-grey {
  color: var(--color-darkest-grey);
}

.g-title--light-grey {
  color: var(--color-light-grey);
}

.g-title--lime-green {
  color: var(--color-lime-green);
}

.g-title--white {
  color: var(--color-white);
}

// Family
.g-title--nunito {
  font-family: 'Nunito';
}

.g-title--seravek {
  font-family: 'Seravek';
}

// Spacing

.g-title--space-normal {
  letter-spacing: normal;
}

.g-title--space-large {
  letter-spacing: 0.65em;
}

// Weight
.g-title--normal {
  font-weight: normal;
}

.g-title--bold {
  font-weight: bold;
}
