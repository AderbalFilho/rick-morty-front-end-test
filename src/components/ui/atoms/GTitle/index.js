import GTitle from './GTitle';

export default GTitle;
