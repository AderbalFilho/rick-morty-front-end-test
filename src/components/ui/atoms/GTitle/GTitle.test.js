import React from 'react';
import TestRenderer from 'react-test-renderer';
import { render } from '@testing-library/react';

import GTitle from './GTitle';

describe('GTitle', () => {
  describe('change color', () => {
    it('renders white when no color is passed', () => {
      const { container } = render(<GTitle>inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--white');
    });

    it('renders darkest-grey', () => {
      const { container } = render(
        <GTitle color="darkest-grey">inGaia</GTitle>
      );
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--darkest-grey');
    });

    it('renders light-grey', () => {
      const { container } = render(<GTitle color="light-grey">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--light-grey');
    });

    it('renders lime-green', () => {
      const { container } = render(<GTitle color="lime-green">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--lime-green');
    });

    it('renders white', () => {
      const { container } = render(<GTitle color="white">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--white');
    });
  });

  describe('change family', () => {
    it('renders Nunito when no family is passed', () => {
      const { container } = render(<GTitle>inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--nunito');
    });

    it('renders Seravek', () => {
      const { container } = render(<GTitle family="seravek">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--seravek');
    });

    it('renders Nunito', () => {
      const { container } = render(<GTitle family="nunito">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--nunito');
    });
  });

  describe('change spacing', () => {
    it('renders normal when no spacing is passed', () => {
      const { container } = render(<GTitle>inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--space-normal');
    });

    it('renders large', () => {
      const { container } = render(<GTitle spacing="large">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--space-large');
    });

    it('renders normal', () => {
      const { container } = render(<GTitle spacing="normal">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--space-normal');
    });
  });

  describe('change tag', () => {
    it('renders h1 when no tag is passed', () => {
      const testRenderer = TestRenderer.create(<GTitle>inGaia</GTitle>);
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders h2', () => {
      const testRenderer = TestRenderer.create(
        <GTitle tag="h2">inGaia</GTitle>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders h3', () => {
      const testRenderer = TestRenderer.create(
        <GTitle tag="h3">inGaia</GTitle>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders h4', () => {
      const testRenderer = TestRenderer.create(
        <GTitle tag="h4">inGaia</GTitle>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders h5', () => {
      const testRenderer = TestRenderer.create(
        <GTitle tag="h5">inGaia</GTitle>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders h6', () => {
      const testRenderer = TestRenderer.create(
        <GTitle tag="h6">inGaia</GTitle>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });

    it('renders h1', () => {
      const testRenderer = TestRenderer.create(
        <GTitle tag="h1">inGaia</GTitle>
      );
      const jsonComponent = testRenderer.toJSON();

      expect(jsonComponent).toMatchSnapshot();
    });
  });

  describe('change weight', () => {
    it('renders normal when no weight is passed', () => {
      const { container } = render(<GTitle>inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--normal');
    });

    it('renders bold', () => {
      const { container } = render(<GTitle weight="bold">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--bold');
    });

    it('renders normal', () => {
      const { container } = render(<GTitle weight="normal">inGaia</GTitle>);
      const title = container.querySelector('.g-title');

      expect(title.classList).toContain('g-title--normal');
    });
  });

  it('renders children', () => {
    const { container } = render(<GTitle>inGaia</GTitle>);
    const title = container.querySelector('.g-title');

    expect(title.textContent).toBe('inGaia');
  });

  it('renders classes', () => {
    const { container } = render(<GTitle classes="rick-title">inGaia</GTitle>);
    const title = container.querySelector('.g-title');

    expect(title.classList).toContain('rick-title');
  });
});
