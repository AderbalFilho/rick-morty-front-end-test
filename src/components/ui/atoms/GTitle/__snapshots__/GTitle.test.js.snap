// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GTitle change tag renders h1 1`] = `
<h1
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h1>
`;

exports[`GTitle change tag renders h1 when no tag is passed 1`] = `
<h1
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h1>
`;

exports[`GTitle change tag renders h2 1`] = `
<h2
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h2>
`;

exports[`GTitle change tag renders h3 1`] = `
<h3
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h3>
`;

exports[`GTitle change tag renders h4 1`] = `
<h4
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h4>
`;

exports[`GTitle change tag renders h5 1`] = `
<h5
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h5>
`;

exports[`GTitle change tag renders h6 1`] = `
<h6
  className="g-title g-title--white g-title--nunito g-title--space-normal g-title--normal"
>
  inGaia
</h6>
`;
